### PR TITLE
JSUI-3260 exchange SAML token in url for access token

### PR DESCRIPTION
**Pull request by samisayegh**
https://coveord.atlassian.net/browse/JSUI-3260

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/ui/AuthenticationProvider/AuthenticationProvider.ts [hunk 1]
```diff
@@ -136,6 +144,56 @@ export class AuthenticationProvider extends Component {
...
+    token && localStorage.setItem(authProviderTemporaryToken, token);
+  }
+
+  private getTemporaryTokenFromUrl() {
+    const fragment = window.location.hash.slice(1);
+    const params = fragment.split('&');
+
+    const tokenParam = _.find(params, param => {
+      const [key] = param.split('=');
+      return key === 'access_token';
```
Comment (dreisch-coveo) [inline src/ui/AuthenticationProvider/AuthenticationProvider.ts:158]: is now `handshake_token`

we try to preserve the rest of hash when adding this but it'll be a good idea to test that out properly as well

**Comment by dreisch-coveo** [comment]
@samisayegh the temporary token is available for 1 minute, and is cleared after the first request for it. do you foresee any issues?

File: src/ui/AuthenticationProvider/AuthenticationProvider.ts (hunk 1 above)
Comment (samisayegh) [reply src/ui/AuthenticationProvider/AuthenticationProvider.ts]: Adjusted the param name, and added a test case for when the token is part of a larger hash string,

**Comment by samisayegh** [comment]
@dreisch-coveo, I do not, one minute should be plenty of time.

File: src/rest/SearchEndpointWithDefaultCallOptions.ts [hunk 2]
```diff
@@ -139,6 +139,10 @@ export class SearchEndpointWithDefaultCallOptions implements ISearchEndpoint {
     return this.endpoint.logError(sentryLog);
   }
 
+  public exchangeAuthenticationProviderHandshakeTokenForAccessToken(token: string): Promise<string> {
```
Comment (olamothe) [inline src/rest/SearchEndpointWithDefaultCallOptions.ts:142]: That's a mouthful ;) 

Suggestion:
```suggestion
  public exchangeAuthenticationProviderToken(token: string): Promise<string> {
```

File: src/ui/AuthenticationProvider/AuthenticationProvider.ts [hunk 3]
```diff
@@ -136,6 +142,51 @@ export class AuthenticationProvider extends Component {
...
+    const promise = this.exchangeHandshakeToken(handshakeToken)
+      .then(token => this.storeAccessToken(token))
+      .then(() => this.loadAccessTokenFromStorage())
+      .catch(e => this.logger.error(e));
+
+    args.defer.push(promise);
+  }
+
+  private exchangeHandshakeToken(token: string) {
+    return SearchEndpoint.defaultEndpoint.exchangeAuthenticationProviderHandshakeTokenForAccessToken(token);
```
Comment (olamothe) [inline src/ui/AuthenticationProvider/AuthenticationProvider.ts:178]: We should not assume `.defaultEndpoint`. 

```suggestion
    return this.queryController.getEndpoint().exchangeAuthenticationProviderHandshakeTokenForAccessToken(token);
```

File: src/ui/AuthenticationProvider/AuthenticationProvider.ts [hunk 4]
```diff
@@ -136,6 +142,51 @@ export class AuthenticationProvider extends Component {
...
+  }
+
+  private storeAccessToken(accessToken: string) {
+    localStorage.setItem(authProviderAccessToken, accessToken);
+  }
+
+  private loadAccessTokenFromStorage() {
+    const token = localStorage.getItem(authProviderAccessToken);
+    token && SearchEndpoint.defaultEndpoint.accessToken.updateToken(token);
```
Comment (olamothe) [inline src/ui/AuthenticationProvider/AuthenticationProvider.ts:187]:
```suggestion
    token && this.queryController.getEndpoint().accessToken.updateToken(token);
```

**Review by olamothe — APPROVED** [review]

**Pull request merged**